trie-db/src/trie_codec.rs

```diff
@@ -0,0 +1,635 @@
+//! perform authenticated lookups on a subset of keys. A naive encoding is the set of encoded nodes
+//! in the partial trie. This, however, includes redundant hashes of other nodes in the partial
+//! trie which could be computed directly. The compact encoding strips out all hash child
+//! references to other nodes in the partial trie and replaces them with empty inline references,
+//! indicating that the child reference is omitted. The nodes are then ordered in pre-order
+//! traversal order so that the full nodes can be efficiently reconstructed recursively. Note that
+//! hash references to nodes not in the partial trie are left intact. The compact encoding can be
+//! expected to save roughly (n - 1) hashes in size where n is the number of nodes in the partial
+//! trie.
+
+use hash_db::HashDB;
+use crate::{
+	CError, ChildReference, DBValue, NibbleVec, NodeCodec, Result,
+	TrieHash, TrieError, TrieDB, TrieDBNodeIterator, TrieLayout,
+	nibble_ops::NIBBLE_LENGTH, node::{Node, NodeHandle, NodeHandlePlan, NodePlan, OwnedNode},
+};
+
+use	::core_::{
+	convert::TryInto,
+	marker::PhantomData,
+	result,
+};
+
+#[cfg(feature = "std")]
+use std::rc::Rc;
+#[cfg(not(feature = "std"))]
+use alloc::rc::Rc;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+struct EncoderStackEntry<C: NodeCodec> {
+	/// The prefix is the nibble path to the node in the trie.
+	prefix: NibbleVec,
+	node: Rc<OwnedNode<DBValue>>,
+	/// The next entry in the stack is a child of the preceding entry at this index. For branch
+	/// nodes, the index is in [0, NIBBLE_LENGTH] and for extension nodes, the index is in [0, 1].
+	child_index: usize,
+	/// Flags indicating whether each child is omitted in the encoded node.
+	omit_children: Vec<bool>,
+	/// The encoding of the subtrie nodes rooted at this entry, which is built up in
+	/// `encode_compact`.
+	encoded_nodes: Vec<Vec<u8>>,
+	_marker: PhantomData<C>,
+}
+
+impl<C: NodeCodec> EncoderStackEntry<C> {
+	/// Given the prefix of the next child node, identify its index and advance `child_index` to
+	/// that. For a given entry, this must be called sequentially only with strictly increasing
+	/// child prefixes. Returns an error if the child prefix is not a child of this entry or if
+	/// called with children out of order.
+	///
+	/// Preconditions:
+	/// - self.prefix + partial must be a prefix of child_prefix.
+	/// - if self.node is a branch, then child_prefix must be longer than self.prefix + partial.
+	fn advance_child_index(&mut self, child_prefix: &NibbleVec)
+		-> result::Result<(), &'static str>
+	{
+		match self.node.node_plan() {
+			NodePlan::Empty | NodePlan::Leaf { .. } =>
+				return Err("empty and leaf nodes have no children"),
+			NodePlan::Extension { .. } => {
+				if self.child_index != 0 {
+					return Err("extension node cannot have multiple children")
+				}
+			}
+			NodePlan::Branch { .. } => {
+				if child_prefix.len() <= self.prefix.len() {
```

Comment:
I wonder if we could just debug_assert that (the function is only call with expect).